README.md
# Taxonomy API for KBase using the Relation Engine

This is an API for taxonomy data (updated from [taxon_api](https://github.com/kbase/taxon_api)) that queries the relation engine.

This service is a registered as a KBase dynamic service under the module name `taxonomy_re_api` and is managed by the Catalog.

## API

JSON RPC requests can be made to to the root path of the URL as a POST request with a json body.

The API follows two conventions followed by KBase APIs:

* All method names are prefixed by the module name (`taxonomy_re_api.get_taxon` instead of just `get_taxon`)
* All params and results are wrapped in an extra array

### Responses

The RPC `"result"` field has the following JSON schema, representing query results from RE:

The result is always wrapped in an array of one element, following KBase convention.

```yaml
type: array
minItems: 1
maxItems: 1
items:
  type: object
  properties:
    total_count:
      type: integer
      title: Result count (before pagination)
    ts:
      type: integer
      title: Timestamp used in the request
    stats:
      type: object
      description: RE query execution meta-info
    results:
      type: array
      items:
        type: object
        additionalProperties: true
        description: Standard RE database fields, plus all additional document-specific fields.
        properties:
          id:
            type: string
            title: RE document ID
            examples: ["562", "287"]
          ns:
            type: string
            title: Taxonomy namespace
            enum: [ncbi_taxonomy]
          created:
            type: integer
            title: Unix epoch creation time
          expired:
            type: integer
            title: Unix epoch expiration time
```

### Timestamp parameter

Every method for this API can take a `ts` parameter, representing the Unix
epoch time (in milliseconds) of when the document was active in the
database. This is optional and defaults to the current time.

## Methods

### taxonomy_re_api.get_taxon(params)

Fetch the document data for a single taxon by ID.

[Request parameters schema (wrapped in an array)](src/server/schemas/get_taxon.yaml)

For the response schema, see the **Responses** section above.

### taxonomy_re_api.get_taxon_from_ws_obj(params)

Fetch the taxon document from a versioned workspace reference.

[Request parameters schema (wrapped in an array)](src/server/schemas/get_taxon_from_ws_obj.yaml)

For the response schema, see the **Responses** section above.

### taxonomy_re_api.get_associated_ws_objects(params)

Fetch all workspace objects associated with a given taxon.

This endpoint is authorization over workspace objects by passing a KBase token via the `Authorization` header.

[Request parameters schema (wrapped in an array)](src/server/schemas/get_associated_ws_objects.yaml)

See the section below about the `select` parameter for further details on it.

For the response schema, see the **Responses** section above.

### taxonomy_re_api.get_lineage(params)

Fetch the ancestors for a taxon vertex.

[Request parameters schema (wrapped in an array)](src/server/schemas/get_lineage.yaml)

See the section below about the `select` parameter for further details on it.

For the response schema, see the **Responses** section above.

### taxonomy_re_api.get_children(params)

Fetch the direct descendants for a taxon vertex.

The optional "search_text" parameter can be specified to search on scientific name for the children. See the **Search text** section below for details on the `search_text` parameter.

[Request parameters schema (wrapped in an array)](/src/server/schemas/get_children.yaml)

See the section below about the `select` parameter for further details on it.

For the response schema, see the **Responses** section above.

### taxonomy_re_api.get_siblings(params)

Fetch the siblings for a taxon.

[Request parameters schema (wrapped in an array)](/src/server/schemas/get_siblings.yaml)

See the section below about the `select` parameter for further details on it.

For the response schema, see the **Responses** section above.

### taxonomy_re_api.search_species(params)

Search for species or strains based on a scientific name. Similar to `search_taxa`, but is a stripped down, faster query.

Paginate with the "limit" and "offset" parameters.

[Request parameters schema (wrapped in an array)](src/server/schemas/search_species.yaml)

See the section below about the `select` parameter for further details on it.

See the **Search text** section below for details on the syntax you can use in the search text.

For the response schema, see the **Responses** section above.

### taxonomy_re_api.search_taxa(params)

Search for taxa based on scientific name.

Paginate with the "limit" and "offset" parameters.

[Request parameters schema (wrapped in an array)](src/server/schemas/search_taxa.yaml)

See the section below about the `select` parameter for further details on it.

See the **Search text** section below for details on the syntax you can use in the search text.

For the response schema, see the **Responses** section above.

### taxonomy_re_api.get_data_sources(params)

Returns all or matching set of taxonomy data source descriptions.

The only parameter is the optional `ns`, which may be used to filter the returned data sources by namespace. There is a 1-1 correspondence between namespaces (e.g. `ncbi_taxonomy`) and a data source descriptions.

#### Params

[Request parameters schema (wrapped in an array)](src/server/schemas/get_data_sources.yaml)
#### Result

Usage:

* omit the params or supply an empty object to return all taxonomy data sources
* supply a list of 1 or more namespaces in the `ns` parameter to return just those data sources

An implication of this design is that supplying an `ns` with an empty list will return no data sources.

#### Example

##### Return all data sources

This RPC request returns all data_sources by omitting the parameter. (Remember, KBase service params are the (almost) always the first element of the JSON-RPC params field, which must be an array.)

```json
{
	"version": "1.1",
	"method": "taxonomy_re_api.get_data_sources",
	"params": []
}
```

or by supplying the params, but omitting the `ns` param:

```json
{
	"version": "1.1",
	"method": "taxonomy_re_api.get_data_sources",
	"params": [{}]
}
```

##### Return NCBI data source

```json
{
	"version": "1.1",
	"method": "taxonomy_re_api.get_data_sources",
	"params": [{"ns": ["ncbi_taxonomy"]}]
}
```

##### Return NCBI and GTDB data source

```json
{
	"version": "1.1",
	"method": "taxonomy_re_api.get_data_sources",
	"params": [{"ns": ["ncbi_taxonomy", "gtdb"]}]
}
```

## Search text

Within a `"search_text"` field, you can use this Arangodb fulltext search syntax to refine the results:

* Separate search terms by comma: "rhodobacter,pseudomonas"
* To search against a term **OR** another term, use a "|" prefix: "rhodobacter,|pseudomonas"
* To search by prefix, use the "prefix:" prefix: "prefix:rhodo,|prefix:pseudo"
* To exclude a term, use `"-term"`

See the ADB documentation for further details: https://www.arangodb.com/docs/stable/aql/functions-fulltext.html

## The `select` parameter

All of the methods (except `get_taxon` and `get_data_sources`) can accept a `select` parameter, which
allows you to control which fields you would like returned. Set it to an array
of the field names that you want. This is useful if you would like to reduce
the size of the response body. If you don't set this parameter, all fields
will be returned in the results.

## Development

### Unit tests

Run unit tests with `make test`

### Integration tests

You can also test the API against a live url. For example:

```
export API_URL="https://ci.kbase.us/dynserv/ecc3fcf41201e66ba6e2d8101195ea29bffba050.taxonomy-re-api"
python -m unittest src/test/integration/test_integration.py

python -m unittest discover -s src/test/integration 
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.9.1] - 2022-05-14
 - Bumped kbase.yml in order to register to beta/release

## [3.9.0] - 2022-04-18
### Changed
 - Removed un-needed GHA actions
 - Upgraded sanic: 19.3.1 -> 20.12.6
 - Upgraded pyyaml: 5.1 -> 5.4

## [3.7.2] - 2022-03-02

### Added

- `get_data_sources` method which returns taxonomy data sources with optional filtering
- `github actions` Docker Image Builds at ghcr.io

### Changed

- extended jsonrpc exceptions to be comprehensive; utilized in main file; extended integration tests to cover.
- `search_species` method now calls the `taxonomy_search_species_strain` and `taxonomy_search_species_strain_no_sort` stored queries in the RE API

## [3.7.1] - 2020-04-14

### Changed

- Use a special RE API URL environment variable if we are in the appdev server environment

## [3.5.0] - 2020-04-02

### Added

- Added the `search_species` method which calls the `taxonomy_search_species` query in the RE API


src/test/test_base.py
import unittest
import os
import requests
import json


def api_url():
    return os.environ.get('API_URL', 'http://localhost:5000')


def verify_ssl():
    return os.environ.get('VERIFY_SSL', 't').lower() not in ['f', 'false', 'n', 'no']


print('Test Settings')
print('-------------')
print(f'API_URL   : {api_url()}')
print(f'VERIFY_SSL: {verify_ssl()}')


TS = 1635479149946  # when test was written


class TestBase(unittest.TestCase):

    # Bundled assertions

    def assert_is_jsonrpc_base(self, rpc_response):
        self.assertIsInstance(rpc_response, dict)
        self.assertIn('version', rpc_response)
        self.assertEqual(rpc_response.get('version'), '1.1')

    def assert_is_error_response(self, rpc_response, code, message, error=None):
        self.assert_is_jsonrpc_base(rpc_response)
        self.assertIn('error', rpc_response)
        rpc_error = rpc_response['error']
        self.assertIsInstance(rpc_error, dict)
        self.assertIsInstance(rpc_error.get('message'), str)
        self.assertEqual(rpc_error.get('code'), code)
        self.assertEqual(rpc_error.get('name'), 'JSONRPCError')
        self.assertEqual(rpc_error.get('message'), message)

        # Test any error detail
        if error is not None:
            self.assertIn('error', rpc_error)
            if 'error' in rpc_error:
                self.assertIsInstance(rpc_error['error'], dict)
                for key, value in error.items():
                    self.assertEqual(rpc_error['error'].get(key), value)
        return error

    def assert_is_result_response(self, rpc_response):
        self.assert_is_jsonrpc_base(rpc_response)
        self.assertIn('result', rpc_response)
        result = rpc_response['result']
        self.assertIsInstance(result, list)
        self.assertEqual(len(result), 1)
        return result[0]

    def assert_query_results(
        self,
        sciname_target,
        search_text_exp_hit,
        search_text_exp_miss=[],
        expired=False,
    ):
        ts_params = {'ts': None} if expired else {}
        for st in search_text_exp_hit + search_text_exp_miss:
            resp = self.request({
                'version': '1.1',
                'method': 'taxonomy_re_api.search_species',
                'params': [{
                    'search_text': st,
                    'ns': 'ncbi_taxonomy',
                    **ts_params,
                }]
            })
            self.assertTrue(resp.ok, resp.text)
            sciname_hits = [
                doc['scientific_name']
                for doc in resp.json()['result'][0]['results']
            ]
            assert_in_method = self.assertIn if st in search_text_exp_hit else self.assertNotIn
            assert_in_method(
                sciname_target,
                sciname_hits,
                f"target: {sciname_target}, query: {st}, sciname_hits: {sciname_hits}"
            )

    def request(self, rpc):
        return requests.post(
            api_url(),
            data=json.dumps(rpc),
            verify=verify_ssl()
        )


src/scripts/run_tests.sh
#!/bin/sh
# set -e

PYTHONPATH=/kb/module

# Run linters
flake8 /kb/module/src
mypy --ignore-missing-imports /kb/module/src
bandit -r /kb/module/src

# Run the server and fork
sh /kb/module/src/scripts/entrypoint.sh &

# Wait for the API to start successfully
python -m src.utils.wait_for_services &&

# Run all tests
python -m pytest -s /kb/module/src/test/unit /kb/module/src/test/integration


src/test/unit/test_misc.py
from src.utils.search import clean_search_text


def test_validate_search_text():
    in_out = [
        ('prefix:', ''),
        ('prefix:e', 'e'),
        ('prefix:esch prefix:col', 'esch prefix:col'),
        ('prefix:~!@#$%^&*()_+', ''),
        ('~!@#$%^&*()_+', ''),
        (', , ,', ''),
        ('Xanthomonas-like sp. V4.BO.41', 'Xanthomonas-like sp. V4.BO.41'),
    ]

    for input, output in in_out:
        assert clean_search_text(input) == output


src/utils/re_api.py
"""
Relation engine API client.
"""
import json
import requests
from src.utils.config import get_config
from src.exceptions import REError

_CONF = get_config()


def query(name, params, tok=None):
    """
    Run a stored query from the RE API.

    Returns (from relation_engine)
    {
        "results": list,        # array of result docs, might be empty
        "count": int,           # num docs
        "has_more": bool,       # whether more results for cursor on server
        "cursor_id": str/None,  # id of temporary cursor on server
        "stats": dict,          # stats
    }

    """
    resp = requests.post(
        _CONF['re_url'] + '/api/v1/query_results',
        params={'stored_query': name},
        data=json.dumps(params),
        headers={'Authorization': tok}
    )
    if not resp.ok:
        raise REError(resp)
    return resp.json()


src/scripts/entrypoint.sh
#!/bin/bash

# Docker container entry point script
# When you run docker run my-app xyz, then this script will get run

# Persistent server mode (aka "dynamic service"):
# This is run when there are no arguments
if [ $# -eq 0 ] ; then
  echo "Running in persistent server mode"
  python -u -m src.server.main

# Test mode
elif [ "${1}" = "test" ] ; then
  echo "Running tests..."
  echo "nothing to do."
  # make test

# Job Mode:
# KBase uses the word "async" for this command, which makes no sense. This is for a one-off job.
elif [ "${1}" = "async" ] ; then
  echo "Running a one-off job..."
  echo "nothing to do."

# Initialize?
elif [ "${1}" = "init" ] ; then
  echo "Initialize module"
  echo "nothing to do."

# This is for registering the module on the KBase catalog
elif [ "${1}" = "report" ] ; then
  echo "Generating report..."
  cp /kb/module/compile_report.json /kb/module/work/compile_report.json

# Unknown command.
else
  echo "Unknown command. Valid commands are: test, async, init, bash, or report"
fi


src/server/main.py
"""
Main HTTP server entrypoint.
"""
import time
import sanic
import jsonschema
import traceback
from jsonschema.exceptions import ValidationError
from contextlib import suppress

from src.utils.config import get_config
from src.utils.schemas import load_schemas
from src.utils.search import clean_search_text
from src.utils import re_api
from src.exceptions import MethodNotFound, InvalidRequest, InvalidParams, ServerError, REError

_CONF = get_config()
_SCHEMAS = load_schemas()
app = sanic.Sanic(name='Taxonomy RE API')
app.config.API_TITLE = 'Taxonomy RE API'
app.config.API_DESCRIPTION = 'Taxonomy data API using the relation engine.'
app.config.API_PRODUCES_CONTENT_TYPES = ['application/json']


def transform_taxon_results(taxa, ns, ns_config):
    """
    Make some modifications on any taxon results given a namespace config (see _NS_CONFIG)
    Mutates each dict in the given `taxa` list
    """
    for taxon in taxa:
        taxon['ns'] = ns


def transform_query_params(params, required_ns_fields=[], field_name_remappings={}):
    """
    Set some defaults in the params that we pass into an RE query.
    `field_name_remappings` can be a dict of {'input_param_field_name': 'field_name_to_send_to_query'}
    Mutates the params dict
    Returns the namespace name and the namespace config dict as a pair (see _NS_CONFIG below)
    """
    ns = params.pop('ns')
    ns_config = _NS_CONFIG[ns]
    for field_name in required_ns_fields:
        params[field_name] = ns_config['query_params'][field_name]
    for (input_name, output_name) in field_name_remappings.items():
        params[output_name] = params.pop(input_name)
    params.setdefault('ts', int(time.time() * 1000))
    return (ns, ns_config)


# Mapping of namespace names to collection and field names
# The 'translate_field_name' options are used to change field names in the source document.
_NS_CONFIG = {
    'ncbi_taxonomy': {
        'query_params': {
            '@taxon_coll': 'ncbi_taxon',
            '@taxon_child_of': 'ncbi_child_of_taxon',
            'sciname_field': 'scientific_name',
        },
    },
    'gtdb': {
        'query_params': {
            '@taxon_coll': 'gtdb_taxon',
            '@taxon_child_of': 'gtdb_child_of_taxon',
            'sciname_field': 'scientific_name',
        },
    },
    'rdp_taxonomy': {
        'query_params': {
            '@taxon_coll': 'rdp_taxon',
            '@taxon_child_of': 'rdp_child_of_taxon',
            'sciname_field': 'name',
        }
    },
    'silva_taxonomy': {
        'query_params': {
            '@taxon_coll': 'silva_taxon',
            '@taxon_child_of': 'silva_child_of_taxon',
            'sciname_field': 'name',
        }
    }
}


def _get_taxon(params, headers):
    """
    Fetch a taxon by ID.
    Returns (result, err), one of which will be None.
    """
    schema = _SCHEMAS['get_taxon']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll',))
    results = re_api.query("taxonomy_fetch_taxon", params)
    transform_taxon_results(results['results'], ns, ns_config)
    return {'stats': results['stats'], 'results': results['results'], 'ts': params['ts']}


def _get_taxon_from_ws_obj(params, headers):
    """
    Fetch the taxon document from a workspace object reference.
    """
    schema = _SCHEMAS['get_taxon_from_ws_obj']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll',))
    params['obj_ref'] = params['obj_ref'].replace('/', ':')
    results = re_api.query("taxonomy_get_taxon_from_ws_obj", params)
    transform_taxon_results(results['results'], ns, ns_config)
    return {'stats': results['stats'], 'results': results['results'], 'ts': params['ts']}


def _get_lineage(params, headers):
    """
    Fetch ancestor lineage for a taxon by ID.
    Returns (result, err), one of which will be None.
    """
    schema = _SCHEMAS['get_lineage']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll', '@taxon_child_of'))
    results = re_api.query("taxonomy_get_lineage", params)
    transform_taxon_results(results['results'], ns, ns_config)
    return {'stats': results['stats'], 'results': results['results'], 'ts': params['ts']}


def _get_children(params, headers):
    """
    Fetch the descendants for a taxon by ID.
    Returns (result, err), one of which will be None.
    """
    schema = _SCHEMAS['get_children']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll', '@taxon_child_of', 'sciname_field'))
    results = re_api.query("taxonomy_get_children", params)
    res = results['results'][0]
    transform_taxon_results(res['results'], ns, ns_config)
    return {'stats': results['stats'], 'total_count': res['total_count'], 'results': res['results'], 'ts': params['ts']}


def _get_siblings(params, headers):
    """
    Fetch the siblings for a taxon by ID.
    Returns (result, err), one of which will be None.
    """
    schema = _SCHEMAS['get_siblings']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll', '@taxon_child_of', 'sciname_field'))
    results = re_api.query("taxonomy_get_siblings", params)
    res = results['results'][0]
    transform_taxon_results(res['results'], ns, ns_config)
    return {'stats': results['stats'], 'total_count': res['total_count'], 'results': res['results'], 'ts': params['ts']}


def _search_taxa(params, headers):
    """
    Search for a taxon vertex by scientific name.
    Returns (result, err), one of which will be None.
    """
    schema = _SCHEMAS['search_taxa']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll', 'sciname_field'))
    results = re_api.query("taxonomy_search_sci_name", params)
    res = results['results'][0]
    transform_taxon_results(res['results'], ns, ns_config)
    return {
        'stats': results['stats'],
        'total_count': res.get('total_count'),
        'results': res['results'],
        'ts': params['ts']
    }


def _search_species(params, headers):
    """
    Search for a species or strain.

    Schema params are:
    * search_text (required)
    * ns (required)
    * ts
    * limit
    * offset
    * select

    Response looks like:
    {
        "results": [...],   # docs
        "count": ...,       # num docs
        "has_more": ...,
        "cursor_id": ...,
        "stats": {...},
    }
    """
    schema = _SCHEMAS['search_species']
    jsonschema.validate(instance=params, schema=schema)
    ns, ns_config = transform_query_params(
        params=params,
        required_ns_fields=('@taxon_coll', 'sciname_field'),
    )
    # Check if the search text is acceptable for AQL
    params['search_text'] = clean_search_text(params['search_text'])
    if params['search_text']:
        stored_query = (
            'taxonomy_search_species_strain_no_sort'
            if len(params['search_text']) <= 3
            else 'taxonomy_search_species_strain'
        )
        resp_json = re_api.query(stored_query, params)
        transform_taxon_results(resp_json['results'], ns, ns_config)
        return {
            'results': resp_json['results'],
            'ts': params['ts'],
            'stats': resp_json['stats'],
        }
    else:
        return {
            'results': [],
            'ts': params['ts'],
            'stats': None,
        }


def _get_associated_ws_objects(params, headers):
    """
    Get any versioned workspace objects associated with a taxon.
    """
    schema = _SCHEMAS['get_associated_ws_objects']
    jsonschema.validate(instance=params, schema=schema)
    (ns, ns_config) = transform_query_params(params, ('@taxon_coll',), {'id': 'taxon_id'})
    results = re_api.query("taxonomy_get_associated_ws_objects", params, headers.get('Authorization'))
    res = results['results'][0]
    transform_taxon_results(results['results'], ns, ns_config)
    for res in results['results']:
        # Write some extra metadata about the workspace for each object
        for elem in res['results']:
            obj = elem['ws_obj']
            obj['workspace'] = {
                'refdata_source': obj['ws_info'].get('metadata', {}).get('refdata_source'),
                'narr_name': obj['ws_info'].get('metadata', {}).get('narrative_nice_Name'),
            }
            del obj['ws_info']
    return {'stats': results['stats'], 'total_count': res['total_count'], 'results': res['results']}


def _get_data_sources(params, headers):
    """
    Returns a list of all Taxonomy Sources
    """
    schema = _SCHEMAS['get_data_sources']

    # parameters for get_data_sources are not actually required, as omitting
    # the parameters (which filter the sources) implies returning all.
    if params is not None:
        jsonschema.validate(instance=params, schema=schema)

    re_params = {
        'type': 'taxonomy',
    }

    # Be nice. Filtering by ns can be skipped by either setting 'ns' to null, or
    # omitting it from the params object.
    if params is not None and params.get('ns') is not None:
        re_params['ns'] = params.get('ns')
        response = re_api.query("data_sources_get_data_sources", re_params, headers.get('Authorization'))
    else:
        response = re_api.query("data_sources_get_all_data_sources", re_params, headers.get('Authorization'))

    sources = []
    for source in response['results']:
        del source['_id']
        del source['_key']
        del source['_rev']
        sources.append(source)

    return {
        'sources': sources
    }


def _rpc_resp(req, resp, status=200):
    resp['version'] = '1.1'
    # We need to suppress the call for json
    # since it fails if the request does not
    # have json.
    has_json = False
    with suppress(Exception):
        if req.json:
            has_json = True

    if has_json and 'id' in req.json:
        resp['id'] = req.json['id']

    return sanic.response.json(resp, status)


def get_json_type(json_value):
    if isinstance(json_value, str):
        return 'string'
    elif isinstance(json_value, bool):
        return 'boolean'
    elif isinstance(json_value, int) or isinstance(json_value, float):
        return 'number'
    elif isinstance(json_value, dict):
        return 'object'
    if isinstance(json_value, list):
        return 'array'
    elif json_value is None:
        return 'null'
    else:
        # should be impossible
        raise Exception('Not a json type')


@app.route('/', methods=["POST", "GET", "OPTIONS"])
async def handle_rpc(req):
    """Handle a JSON RPC 1.1 request."""
    if req.method == 'OPTIONS':
        return sanic.response.raw(b'', status=204)
    if req.method == 'GET':
        # Server status request
        return _rpc_resp(req, {'result': [{'status': 'ok'}]})
    body = req.json
    handlers = {
        'taxonomy_re_api.get_taxon': _get_taxon,
        'taxonomy_re_api.get_lineage': _get_lineage,
        'taxonomy_re_api.get_children': _get_children,
        'taxonomy_re_api.get_siblings': _get_siblings,
        'taxonomy_re_api.search_taxa': _search_taxa,
        'taxonomy_re_api.search_species': _search_species,
        'taxonomy_re_api.get_associated_ws_objects': _get_associated_ws_objects,
        'taxonomy_re_api.get_taxon_from_ws_obj': _get_taxon_from_ws_obj,
        'taxonomy_re_api.get_data_sources': _get_data_sources,
    }

    # Validate  JSON-RPC 1.1 overall structure

    if not body:
        raise InvalidRequest("Request is not valid")

    if not isinstance(body, dict):
        raise InvalidRequest("Request is not valid")

    # Validate version

    if 'version' not in body:
        raise InvalidRequest("Missing version version in request")

    if body['version'] != '1.1':
        raise InvalidRequest("Version must be 1.1")

    # Validate method

    if 'method' not in body:
        raise InvalidRequest("Missing method name in request")

    method = body['method']

    if not isinstance(method, str):
        raise InvalidRequest("Request method is invalid")

    # Validate params

    if 'params' not in body:
        raise InvalidRequest("Missing params in request")

    params = body['params']

    if not isinstance(params, list):
        raise InvalidRequest(f'Method params should be an array, it is a "{get_json_type(params)}"')

    param = None
    if len(params) == 1:
        param = params[0]
    elif len(params) > 1:
        raise InvalidParams(f"Method params array can only include at most one item, it has {len(params)}")

    # Run the method
    if method not in handlers:
        raise MethodNotFound(method)

    meth = handlers[method]

    result = meth(param, req.headers)
    resp = {'result': [result]}
    return _rpc_resp(req, resp)


@app.middleware('response')
async def cors_resp(req, res):
    """Handle cors response headers."""
    res.headers['Access-Control-Allow-Origin'] = '*'
    res.headers['Access-Control-Allow-Methods'] = 'POST, GET, OPTIONS'
    res.headers['Access-Control-Allow-Headers'] = '*'


@app.exception(sanic.exceptions.NotFound)
async def page_not_found(req, err):
    """Handle 404 as a json response."""
    resp = {
        'error': {
            'name': 'not_found',
            'code': 404,
            'message': 'Not found - ' + str(err),
        }
    }
    return _rpc_resp(req, resp, status=404)


@app.exception(ValidationError)
async def invalid_schema(req, err):
    """Handle a JSON Schema validation error."""
    error = {
        'message': 'Parameter validation error: ' + err.message,
        'validator': err.validator,
        'validator_value': err.validator_value,
        'path': list(err.path),
    }
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': InvalidParams.code,
            'message': 'Invalid params',
            'error': error
        }
    }
    return _rpc_resp(req, resp, status=400)


@app.exception(REError)
async def re_api_error(req, err):
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': ServerError.code,
            'message': 'Server error',
            'error': {
                'message': 'Relation engine API error',
                're_error': err.resp_json or err.resp_text
            }
        }
    }
    return _rpc_resp(req, resp, status=400)


@app.exception(sanic.exceptions.InvalidUsage)
async def invalid_usage(req, err):
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': InvalidRequest.code,
            'message': 'Invalid request',
            'error': {
                'message': str(err),
            }

        }
    }
    return _rpc_resp(req, resp, status=400)


@app.exception(sanic.exceptions.MethodNotSupported)
async def method_not_supported(req, err):
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': InvalidRequest.code,
            'message': 'Invalid request',
            'error': {
                'message': str(err),
            }
        }
    }
    return _rpc_resp(req, resp, status=400)


@app.exception(InvalidRequest)
async def invalid_request(req, err):
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': err.code,
            'message': 'Invalid request',
            'error': {
                'message': str(err)
            }
        }
    }
    return _rpc_resp(req, resp, status=400)


@app.exception(MethodNotFound)
async def method_not_found(req, err):
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': err.code,
            'message': 'Method not found',
            'error': {
                'message': f'Method "{err.method_name}" not found',
                'method': err.method_name
            }
        }
    }
    return _rpc_resp(req, resp, status=400)


@app.exception(InvalidParams)
async def invalid_params(req, err):
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': err.code,
            'message': 'Invalid params',
            'error': {
                'message': str(err),
            }
        }
    }
    return _rpc_resp(req, resp, status=400)


# Any other exception -> 500
@app.exception(Exception)
async def server_error(req, err):
    traceback.print_exc()
    resp = {
        'error': {
            'name': 'JSONRPCError',
            'code': ServerError.code,
            'message': 'Server error',
            'error': {
                'message': str(err),
                'class': err.__class__.__name__
            }
        }
    }
    return _rpc_resp(req, resp, status=500)

if __name__ == '__main__':
    app.run(
        host='0.0.0.0',  # nosec
        port=5000,
        debug=_CONF['dev'],
        workers=_CONF['nworkers']
    )
